### PR TITLE
feat(python): support `DataFrame` init from queries against user-instantiated database connections

### PR DESCRIPTION
**Pull request by alexander-beedie**
Closes: #10121.

Ref: #10416, #10111, #10177.

## Overview

This is a massive expansion of our current database support, but requires no new dependencies. It allows users to pass all manner of instantiated connections, representing a huge number of disparate backends, loading the data as Arrow where possible, otherwise falling back to row-wise init.

## Supported connection types

* ADBC
* Databricks
* SQLAlchemy
* PyODBC
* Turbodbc
* DuckDB
* Snowflake connector
* Anything DBAPI2 compatible

All the differences are handled automagically by the new `ConnectionExecutor`, which takes care of all the minor variations in cursor behaviour (eg: in-place execution vs returning a separate resultset object), whether or not the driver is Arrow-aware, etc...

## Loading as Arrow

Registering Arrow resultset awareness is as simple as making a trivial entry into the private `_ARROW_DRIVER_REGISTRY_` dictionary - for example:

```python
"databricks": {
    "fetch_all": "fetchall_arrow",
    "fetch_batches": "fetchmany_arrow",
    "exact_batch_size": True,
},
```
Once registered, these methods are favoured on a per-driver basis over their row-wise `fetchall` and `fetchmany` equivalents. Currently ADBC, Databricks, Snowflake, Turbodcbc and DuckDB have all been registered as Arrow-aware. If anyone knows any others then let's add them.

## String URI migration

The `read_database` function has now been split; for the time being it will continue to accept string URIs, raising a DeprecationWarning and redirecting to the new `read_database_uri` function if given one.

After a suitable transition period we can start raising an error instead of a warning when string URIs are passed to `read_database`.

## Testing

Obviously we don't have access to all of these databases for testing, so we may find a few small issues when users get hold of it; however, I have tried it out with as many as I could get my hands on and I think it's looking to be in pretty decent shape... 

Have also mocked some Connection objects to help validate that we _are_ actually calling the correct Arrow methods for those drivers that have Arrow support .

## Future additions

* Arrow FlightSQL clients such as Influx/Dremio.
* Google BigQuery Client connection objects.

(I have no access to Influx/Dremio backends, so these will be a bit tricky to validate).

## Example

Create some test data in a SQLite db file:
```python
import sqlite3
from sqlalchemy import create_engine
from adbc_driver_sqlite.dbapi import connect as adbc_connect

test_db = "test_data.db"
sqlite_conn = sqlite3.connect( test_db )
sqlite_conn.executescript(
    """
    CREATE TABLE test_data (id INT PRIMARY KEY, key TEXT, value FLOAT);
    INSERT INTO test_data(key,value) VALUES ('misc',100.0), ('other',-99.5);
    """
)
```
Read the data into a polars DataFrame using three completely different flavours of SQLite driver, using an identical call to `read_database`.
```python
for conn in (
    sqlite_conn,  # native sqlite3
    adbc_connect(test_db),  # adbc (arrow)
    create_engine(f"sqlite:///{test_db}"),  # sqlalchemy
):
    pl.read_database(
        query="SELECT * FROM test_data ORDER BY value",
        connection=conn,
    )
    # shape: (2, 3)
    # ┌─────┬───────┬───────┐
    # │ id  ┆ key   ┆ value │
    # │ --- ┆ ---   ┆ ---   │
    # │ i64 ┆ str   ┆ f64   │
    # ╞═════╪═══════╪═══════╡
    # │ 2   ┆ other ┆ -99.5 │
    # │ 1   ┆ misc  ┆ 100.0 │
    # └─────┴───────┴───────┘

```

### CODE REVIEW

File: py-polars/polars/io/database.py [hunk 1]
```diff
@@ -3,20 +3,265 @@
...
+        "exact_batch_size": False,
+    },
+}
+
+
+class ConnectionExecutor:
+    """Abstraction for querying databases with user-supplied connection objects."""
+
+    def __init__(self, connection: ConnectionOrCursor) -> None:
+        self.driver = type(connection).__module__.split(".", 1)[0].lower()
```
[the remark this answers is not in the document]
Comment (alexander-beedie) [reply py-polars/polars/io/database.py:63]: No; it'll just fall-through to basic row-wise execution if we don't have it registered as being Arrow-aware, and if it still can't execute in that context then it'll raise a nice clear error. (I should add an extra unit test to validate that).

[the remark this answers is not in the document]
Comment (ritchie46) [reply py-polars/polars/io/database.py:63]: ah, TIL about dbapi2. :+1:

**Review by ritchie46 — APPROVED** [review]

**Comment by stinodego** [comment]
Could you explain to me exactly what the difference is between `read_database` and `read_database_uri`? The original `read_database` already took a connection URI, right? Why do we need a separate function for this new functionality?

**Comment by alexander-beedie** [comment]
> This is great work 🎉

✌️😎   

> One question though: could you explain to me exactly what the difference is between `read_database` and `read_database_uri`? The original `read_database` already took a connection URI, right? Why do we need a separate function for this new functionality?

I actually started out on that track, but very quickly found that it made for an ugly API and messier internals; for example, _none_ of the following parameters make any sense if you start out with a connection:

* partition_on
* partition_range
* partition_num
* protocol
* engine

(They also don't make any sense for `adbc`, but that may be a cleanup for another day ;)

At this point the URI function is really _only_ useful for `connectorx` (as you can create `adbc` connections yourself); separating the two means that we have a much cleaner and more-focused API for `read_database` (which is good as I suspect that user-supplied connections will end up being the dominant calling pattern), and we can evolve it as needed _specifically_ for use with connections without dragging around a lot of `connectorx` or URI parameter baggage. Internally it also keeps things cleaner with respect to connection management/creation. 

So really boils down to having a clean API (as the user can setup their backend-specific connection params without us having to get involved in parameter selection) and more obvious behaviour/expectations as regards connection management.

File: py-polars/pyproject.toml [hunk 2]
```diff
@@ -164,6 +164,7 @@ strict = true
 
 [tool.ruff.per-file-ignores]
 "tests/**/*.py" = ["D100", "D103", "B018"]
+"polars/type_aliases.py" = ["D101", "D102"]
```
Comment (stinodego) [inline py-polars/pyproject.toml:167]: Please just use `# noqa` comments for these.

Or even better, write a small docstring.

**Review by stinodego — COMMENTED** [review]
Okay, I'm sold!

I just have 1 minor nitpick, otherwise looks great and can be merged!

File: py-polars/polars/type_aliases.py [hunk 3]
```diff
@@ -193,3 +194,32 @@
 # typevars for core polars types
 PolarsType = TypeVar("PolarsType", "DataFrame", "LazyFrame", "Series", "Expr")
 FrameType = TypeVar("FrameType", "DataFrame", "LazyFrame")
+
+
+# minimal protocol objects that can reasonably represent
+# an executable connection, cursor, or equivalent object
+class BasicConnection(Protocol):
```
[the remark this answers is not in the document]
Comment (alexander-beedie) [reply py-polars/polars/type_aliases.py:201]: Literally the first one I've ever written, lol... but exactly the right tool for the job in this case :))

File: py-polars/pyproject.toml (hunk 2 above)
Comment (alexander-beedie) [reply py-polars/pyproject.toml]: Lol... done  ;)

**Review by stinodego — APPROVED** [review]

**Comment by jonashaag** [comment]
This is awesome!

@alexander-beedie, quoting you from [here](https://github.com/pola-rs/polars/issues/10121#issuecomment-1654530017):

> This "adaptor" is likely to be trivial

Do you still consider this adapter to be trivial? I still wonder if it's worth having those 500 lines of code in a separate library.

Relatedly, would you be interested in co-authoring/feedbacking on an extension to [PEP 249](https://peps.python.org/pep-0249/)?

**Comment by alexander-beedie** [comment]
> Do you still consider this adapter to be trivial?

Probably a bit less trivial than originally anticipated, indeed ;)

> I still wonder if it's worth having those 500 lines of code in a separate library.

Perhaps, but it's not so large that I think we should take an external dependency on it, when it's currently tuned to being exactly what we need. I'm not averse to making a separate standalone version though - might be a useful little project! 🤔

> Relatedly, would you be interested in co-authoring/feedbacking on an extension to [PEP 249](https://peps.python.org/pep-0249/)?

DM me with thoughts/details on the Discord? Sounds interesting to me... :)

**Comment by mcrumiller** [comment]
@alexander-beedie are there plans to have `write_database` allow existing connections as well? Currently with polars I cannot use temporary tables since they require a persistent connection, which would be pretty useful in my case.

**Comment by alexander-beedie** [comment]
> @alexander-beedie are there plans to have `write_database` allow existing connections as well?

Yes, but it's a lot more work than `read` to do well.

**Comment by ldacey** [comment]
I have been using a library to retrieve data with ODBC into pyarrow tables (after challenges with installing turbodbc with pip instead of conda): https://github.com/pacman82/arrow-odbc-py/blob/main/python/arrow_odbc/connect.py

It seems like the connection object doesn't work. Does something need to be changed in the arrow_odbc library to enable support with pl.read_database? Just wondering where I should raise the issue.

```
TypeError: Unrecognised connection <cdata 'struct OdbcConnection *' 0x558eb967f6a0>; unable to find 'execute' method
```

**Comment by alexander-beedie** [comment]
> It seems like the connection object doesn't work. 

Most of the Arrow-supporting libraries have non-standard method names for returning result sets, so we need to specifically register them.

> Does something need to be changed in the arrow_odbc library to enable support with pl.read_database? Just wondering where I should raise the issue.

Raise a feature request Issue for this library and I'll look into it today; shouldn't be hard to add support 👍

**Pull request merged**